remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java

```diff
@@ -0,0 +1,215 @@
+    public InputStream fetchOffsetIndex(final RemoteLogSegmentMetadata remoteLogSegmentMetadata) throws RemoteStorageException {
+        Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentMetadata must not be null");
+
+        final S3RemoteLogSegmentContext context = deserializeS3RemoteLogSegmentContext(remoteLogSegmentMetadata.remoteLogSegmentContext());
+        final String offsetIndexFileKey = s3Key(remoteLogSegmentMetadata.remoteLogSegmentId(), context.offsetIndexFileName());
+
+        try {
+            final S3Object s3Object = s3Client.getObject(bucket, offsetIndexFileKey);
+            return s3Object.getObjectContent();
+        } catch (final SdkClientException e) {
+            throw new RemoteStorageException("Error fetching offset index from " + offsetIndexFileKey, e);
+        }
+    }
+
+    @Override
+    public InputStream fetchTimestampIndex(final RemoteLogSegmentMetadata remoteLogSegmentMetadata) throws RemoteStorageException {
+        Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentMetadata must not be null");
+
+        final S3RemoteLogSegmentContext context = deserializeS3RemoteLogSegmentContext(remoteLogSegmentMetadata.remoteLogSegmentContext());
+        final String timeIndexFileKey = s3Key(remoteLogSegmentMetadata.remoteLogSegmentId(), context.timeIndexFileName());
+
+        try {
+            final S3Object s3Object = s3Client.getObject(bucket, timeIndexFileKey);
+            return s3Object.getObjectContent();
+        } catch (final SdkClientException e) {
+            throw new RemoteStorageException("Error fetching timestamp index from " + timeIndexFileKey, e);
+        }
+    }
+
+    @Override
+    public void deleteLogSegment(final RemoteLogSegmentMetadata remoteLogSegmentMetadata) throws RemoteStorageException {
+        Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentMetadata must not be null");
+
+        final S3RemoteLogSegmentContext context = deserializeS3RemoteLogSegmentContext(remoteLogSegmentMetadata.remoteLogSegmentContext());
+        final String logFileKey = s3Key(remoteLogSegmentMetadata.remoteLogSegmentId(), context.logFileName());
+        final String offsetIndexFileKey = s3Key(remoteLogSegmentMetadata.remoteLogSegmentId(), context.offsetIndexFileName());
+        final String timeIndexFileKey = s3Key(remoteLogSegmentMetadata.remoteLogSegmentId(), context.timeIndexFileName());
+
+        try {
+            final DeleteObjectsRequest deleteObjectsRequest = new DeleteObjectsRequest(bucket)
+                .withKeys(logFileKey, offsetIndexFileKey, timeIndexFileKey);
+            s3Client.deleteObjects(deleteObjectsRequest);
+        } catch (final SdkClientException e) {
+            throw new RemoteStorageException(String.format("Error deleting %s, %s or %s", logFileKey, offsetIndexFileKey, timeIndexFileKey), e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        // intentionally left blank
```

Comment:
Don't we need to close S3 client (not sure it is offered by the API?)

Reply:
Yes, sure